From 6257915d3c5b9323bd2277ceb9e24021d2ad1179 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Wed, 11 Sep 2024 17:45:41 +0300
Subject: [PATCH] [FSW-8694] fs1: write slave spi fuse values

With this patch added support in master die to
read all slave die's spi fuse values and write
them into its scratch_pad_0/1, and inform to
slave by writing magic into its scratch_pad_2
register.

Change-Id: I0252e47cddbd89ca2d7e33b65b2d692010da9980
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 340fe13..2430f6b 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -29,8 +29,65 @@
 #include <gpio_regs.h>
 #include <psoc_boot_conf_masks.h>
 #include <arc_regs.h>
+#include <boot_desc.h>
+#include <dev_spi_flash.h>
 
 extern const int fs1_die_type;
+extern uint32_t fake_fuse_offset_array[MAX_DIE_COUNT];
+
+void boot_rom_init_slave_spi_efuse()
+{
+	/* Align with enum fake_fuse_bin_index */
+	uint8_t die_id_arary[MAX_DIE_COUNT] = {FS1_IO_DIE_0, FS1_IO_DIE_1,
+		FS1_COMPUTE_DIE_0, FS1_COMPUTE_DIE_1, FS1_COMPUTE_DIE_2, FS1_COMPUTE_DIE_3};
+	uint32_t bank_offset, spi_offset, syscfg1_data, syscfg2_data;
+	uint8_t die_id_index;
+
+	/* Note: Index 0 master, no need to export. */
+	for(die_id_index = 1 ; die_id_index < MAX_DIE_COUNT; die_id_index++) {
+
+		spi_offset = fake_fuse_offset_array[die_id_index];
+
+		if (spi_offset == 0)
+			continue;
+
+		/* read value from bank0 */
+		bank_offset = (IO_EFUSE_SYSCFG_OFFSET * sizeof(uint32_t));
+		syscfg1_data = spi_flash_readl(spi_offset + bank_offset);
+
+		/* read value from bank1 and OR with value from bank0 */
+		bank_offset = ((EFUSE_WORDS_PER_BANK + IO_EFUSE_SYSCFG_OFFSET) * sizeof(uint32_t));
+		syscfg1_data |= spi_flash_readl(spi_offset + bank_offset);
+
+		/* read value from bank0 */
+		bank_offset = (IO_EFUSE_SYSCFG2_OFFSET * sizeof(uint32_t));
+		syscfg2_data = spi_flash_readl(spi_offset + bank_offset);
+
+		/* read value from bank1 and OR with value from bank0 */
+		bank_offset = ((EFUSE_WORDS_PER_BANK + IO_EFUSE_SYSCFG2_OFFSET) * sizeof(uint32_t));
+		syscfg2_data |= spi_flash_readl(spi_offset + bank_offset);
+
+		/* Notify slave config fuse value in slave scratchpad 0/1 register */
+		regmap_writel_inter_die(
+			IO0_H0_PSOC_ARC_MNG_LOCAL(mmARC_LOCAL_CONF_CFG_ADDR_MNPL_CTRL),
+			syscfg1_data,
+			DIE_BASED_ADDRESS((IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_SCRATCHPAD_0)),
+				die_id_arary[die_id_index]));
+
+		regmap_writel_inter_die(
+			IO0_H0_PSOC_ARC_MNG_LOCAL(mmARC_LOCAL_CONF_CFG_ADDR_MNPL_CTRL),
+			syscfg2_data,
+			DIE_BASED_ADDRESS((IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_SCRATCHPAD_1)),
+				die_id_arary[die_id_index]));
+
+		/* Indicate slave die about fuse value is available */
+		regmap_writel_inter_die(
+			IO0_H0_PSOC_ARC_MNG_LOCAL(mmARC_LOCAL_CONF_CFG_ADDR_MNPL_CTRL),
+			FFUSE_WRITE_IND_MAGIC,
+			DIE_BASED_ADDRESS((IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_SCRATCHPAD_2)),
+				die_id_arary[die_id_index]));
+	}
+}
 
 uint8_t boot_rom_get_die_id(void)
 {
diff --git a/common/common.c b/common/common.c
index fda1ffe..e66f180 100644
--- a/common/common.c
+++ b/common/common.c
@@ -146,7 +146,8 @@
 
 	boot_rom_init_sku();
 
-	/* TODO: Send fuse values to slaves */
+	if ((boot_rom_global.sku == FS1_FULL_PACKAGE) && boot_rom_in_master_die())
+		boot_rom_init_slave_spi_efuse();
 
 	boot_rom_set_spi_efuse_read();
 }
diff --git a/common/common.h b/common/common.h
index b0e7491..4d591be 100644
--- a/common/common.h
+++ b/common/common.h
@@ -38,5 +38,6 @@
 void boot_rom_start_cpld_wdtimer(void);
 bool boot_rom_spi_efuse_enabled(void);
 void boot_rom_config_arc_regions(void);
+void boot_rom_init_slave_spi_efuse(void);
 
 #endif /* _COMMON_H_ */
diff --git a/common/spi_flash.h b/common/spi_flash.h
index e3688b0..793e537 100644
--- a/common/spi_flash.h
+++ b/common/spi_flash.h
@@ -23,6 +23,9 @@
 
 #define SPI_FLASH_BAUDR_DIVISOR		2
 
+/* Fuse emulation indication magic:	  F I N D (Fuse INDication) */
+#define FFUSE_WRITE_IND_MAGIC		0x46494E44
+
 void spi_flash_init(void);
 uint32_t spi_flash_fuse_read(uint32_t word, uint32_t *val);
 
diff --git a/include/mem_io.h b/include/mem_io.h
index 1957355..d5e9bd5 100644
--- a/include/mem_io.h
+++ b/include/mem_io.h
@@ -23,6 +23,9 @@
 #define BIT(nr)					(1 << (nr))
 #define MASK_32_BIT				0xFFFFFFFF
 
+#define DIE_BASED_ADDRESS(offset, die_id)	(offset & ~(HL_ARC_ADDR_DIE_ID_MASK)) | \
+				(die_id << HL_ARC_ADDR_DIE_ID_SHIFT)
+
 static inline void base_writel(uint32_t val, uint32_t offset)
 {
 	*((volatile uint32_t *)(offset)) = val;
